clojure/src/ch01_evaluator.clj

```diff
@@ -0,0 +1,112 @@
+(ns ch01-evaluator
+  "Chapter 1: Basic evaluator.
+  Starts with section 1.2 on p. 4.")
+
+;;; 1.2 (p.4): define `evaluate` signature
+(defn evaluate [exp env])
+
+
+;;; 1.3 (p.4): start building the implementation
+;;; - introduces `atom?` check and `case` for non-atomic forms
+;;; - discusses the difference between 'program' and 'representation';
+;;;   evaluate works on the _representation_ 
+
+;; they define atom as anything that's not a "pair" (cons cell)
+;; - we use `list?` for Clojure which should be good enough
+(def atom? (complement list?))
+(atom? 1)
+;; => true
+(atom? "ahoj")
+;; => true
+(atom? \c)
+;; => true
+(atom? '(1 2 3))
+;; => false
+;; TODO: notice that other composite data structures like vectors are considered atoms!
+(atom? [1 2 3])
+;; => true
+(atom? {1 2 3 4})
+;; => true
+
+
+(defn evaluate [exp env]
+  (if (atom? exp)
+    ,,,
+    ;; we use `first` instead of `car`
+    (case (first exp)
+      ,,,)))
```

Comment:
I had no idea that you can sort of iteratively build up function definitions like this? I guess that successive calls to `defn` simply redefine the symbol?

Reply:
This is cool btw! Makes it super nice to read.

Reply:
Yes, exactly - it's typically an error to have duplicate definitions but for a nice "trick" to use for building up the stuff incrementally.